Bug: no-invalid-properties reports "Unknown property" when the value contains env()

### Environment

ESLint version: v10.7.0
@eslint/css version: v1.4.0
Node version: v22.15.0
npm version: v11.12.1
Operating System: Windows 11


### Which language are you using?

stylesheet

### What did you do?

<details>
<summary>Configuration</summary>

```js
import css from "@eslint/css";

export default [
	{
		files: ["**/*.css"],
		plugins: { css },
		language: "css/css",
		rules: {
			"css/no-invalid-properties": "error",
		},
	},
];
```
</details>

```css
.a {
	padding-top: env(safe-area-inset-top);
}

.b {
	padding-top: calc(env(safe-area-inset-top) + 10px);
}
```

### What did you expect to happen?

No errors. These are all valid declarations — env() is standard CSS and padding-top/width are obviously real properties.

### What actually happened?

Both declarations are reported as an unknown property:

```
  2:2  error  Unknown property 'padding-top' found  css/no-invalid-properties
  6:2  error  Unknown property 'padding-top' found  css/no-invalid-properties
```

css-tree's lexer throws a plain Error ("Matching for a tree with env() is not supported") when the value contains env(), and the error handling in no-invalid-properties treats any non-SyntaxMatchError as an unknown property when allowUnknownVariables is false. So the "env() not supported" error ends up reported as "Unknown property".

Setting allowUnknownVariables: true works around it, but the default hits pretty common CSS (safe-area-inset padding) and the message is wrong either way. Maybe values containing env() should just skip validation, like
unresolved var() does?

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-2hdjwmbx

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

_No response_

--- a/src/rules/no-invalid-properties.js
+++ b/src/rules/no-invalid-properties.js
@@ -7,7 +7,11 @@
 // Imports
 //-----------------------------------------------------------------------------
 
-import { isSyntaxMatchError, isSyntaxReferenceError } from "../util.js";
+import {
+	isSyntaxMatchError,
+	isSyntaxReferenceError,
+	isEnvMatchError,
+} from "../util.js";
 
 //-----------------------------------------------------------------------------
 // Type Definitions
@@ -463,7 +467,13 @@ export default /** @satisfies {NoInvalidPropertiesRuleDefinition} */ ({
 						});
 						return;
 					}
-
+					if (isEnvMatchError(error)) {
+						/*
+						 * env() values are provided by the user agent and
+						 * cannot be validated, so skip validation entirely.
+						 */
+						return;
+					}
 					if (
 						!allowUnknownVariables ||
 						isSyntaxReferenceError(error)

--- a/src/util.js
+++ b/src/util.js
@@ -32,3 +32,13 @@ export function isSyntaxMatchError(error) {
 export function isSyntaxReferenceError(error) {
 	return typeof error.reference === "string";
 }
+
+/**
+ * Determines if an error is the lexer error thrown when a value containing
+ * `env()` cannot be matched against a syntax definition.
+ * @param {Object} error The error object to check.
+ * @returns {boolean} True if the error is the `env()` match error, false if not.
+ */
+export function isEnvMatchError(error) {
+	return error.message === "Matching for a tree with env() is not supported";
+}

--- a/tests/rules/no-invalid-properties.test.js
+++ b/tests/rules/no-invalid-properties.test.js
@@ -30,6 +30,7 @@ ruleTester.run("no-invalid-properties", rule, {
 		"a { padding-top: calc(env(safe-area-inset-top) + 10px); }",
 		"a { width: env(titlebar-area-width); }",
 		"a { padding: env(safe-area-inset-top) 0 env(safe-area-inset-bottom) 0; }",
+		"a { padding: env(safe-area-inset-top) red }",
 		"a { color: red; background-color: blue; }",
 		"a { color: red; transition: none; }",
 		"body { --custom-property: red; }",

--- a/tests/rules/no-invalid-properties.test.js
+++ b/tests/rules/no-invalid-properties.test.js
@@ -25,6 +25,11 @@ const ruleTester = new RuleTester({
 ruleTester.run("no-invalid-properties", rule, {
 	valid: [
 		"a { color: red; }",
+		"a { padding-top: env(safe-area-inset-top); }",
+		"a { padding-top: env(safe-area-inset-top, 20px); }",
+		"a { padding-top: calc(env(safe-area-inset-top) + 10px); }",
+		"a { width: env(titlebar-area-width); }",
+		"a { padding: env(safe-area-inset-top) 0 env(safe-area-inset-bottom) 0; }",
 		"a { color: red; background-color: blue; }",
 		"a { color: red; transition: none; }",
 		"body { --custom-property: red; }",
@@ -276,6 +281,19 @@ ruleTester.run("no-invalid-properties", rule, {
 		"main { p:first-of-type, span { color: red; } }",
 	],
 	invalid: [
+		{
+			code: "a { paddin-top: env(safe-area-inset-top); }",
+			errors: [
+				{
+					messageId: "unknownProperty",
+					data: { property: "paddin-top" },
+					line: 1,
+					column: 5,
+					endLine: 1,
+					endColumn: 15,
+				},
+			],
+		},
 		{
 			code: "a { color: bar }",
 			errors: [

--- a/tests/rules/no-invalid-properties.test.js
+++ b/tests/rules/no-invalid-properties.test.js
@@ -25,12 +25,6 @@ const ruleTester = new RuleTester({
 ruleTester.run("no-invalid-properties", rule, {
 	valid: [
 		"a { color: red; }",
-		"a { padding-top: env(safe-area-inset-top); }",
-		"a { padding-top: env(safe-area-inset-top, 20px); }",
-		"a { padding-top: calc(env(safe-area-inset-top) + 10px); }",
-		"a { width: env(titlebar-area-width); }",
-		"a { padding: env(safe-area-inset-top) 0 env(safe-area-inset-bottom) 0; }",
-		"a { padding: env(safe-area-inset-top) red }",
 		"a { color: red; background-color: blue; }",
 		"a { color: red; transition: none; }",
 		"body { --custom-property: red; }",
@@ -280,41 +274,16 @@ ruleTester.run("no-invalid-properties", rule, {
 		"main { p:has(.child) { color: red; } }",
 		"main { p:has(.child:hover) { color: red; } }",
 		"main { p:first-of-type, span { color: red; } }",
+
+		// env() values are provided by the user agent and cannot be validated
+		"a { padding-top: env(safe-area-inset-top); }",
+		"a { padding-top: env(safe-area-inset-top, 20px); }",
+		"a { padding-top: calc(env(safe-area-inset-top) + 10px); }",
+		"a { width: env(titlebar-area-width); }",
+		"a { padding: env(safe-area-inset-top) 0 env(safe-area-inset-bottom) 0; }",
+		"a { padding: env(safe-area-inset-top) red }",
 	],
 	invalid: [
-		{
-			code: "a { paddin-top: env(safe-area-inset-top); }",
-			errors: [
-				{
-					messageId: "unknownProperty",
-					data: { property: "paddin-top" },
-					line: 1,
-					column: 5,
-					endLine: 1,
-					endColumn: 15,
-				},
-			],
-		},
-
-		/*
-		 * Reporting an unknown `var()` nested inside `env()` is intentional:
-		 * this rule also checks that custom properties are resolvable, and that
-		 * check is independent of the syntax validation skipped for `env()`.
-		 * We can revisit this once the rule supports partial validation of values.
-		 */
-		{
-			code: "a { padding: env(safe-area-inset-top, var(--external-padding)); }",
-			errors: [
-				{
-					messageId: "unknownVar",
-					data: { var: "--external-padding" },
-					line: 1,
-					column: 43,
-					endLine: 1,
-					endColumn: 61,
-				},
-			],
-		},
 		{
 			code: "a { color: bar }",
 			errors: [
@@ -1490,6 +1459,39 @@ ruleTester.run("no-invalid-properties", rule, {
 					column: 56,
 					endLine: 1,
 					endColumn: 63,
+				},
+			],
+		},
+		{
+			code: "a { paddin-top: env(safe-area-inset-top); }",
+			errors: [
+				{
+					messageId: "unknownProperty",
+					data: { property: "paddin-top" },
+					line: 1,
+					column: 5,
+					endLine: 1,
+					endColumn: 15,
+				},
+			],
+		},
+
+		/*
+		 * Reporting an unknown `var()` nested inside `env()` is intentional:
+		 * this rule also checks that custom properties are resolvable, and that
+		 * check is independent of the syntax validation skipped for `env()`.
+		 * We can revisit this once the rule supports partial validation of values.
+		 */
+		{
+			code: "a { padding: env(safe-area-inset-top, var(--external-padding)); }",
+			errors: [
+				{
+					messageId: "unknownVar",
+					data: { var: "--external-padding" },
+					line: 1,
+					column: 43,
+					endLine: 1,
+					endColumn: 61,
 				},
 			],
 		},

--- a/tests/rules/no-invalid-properties.test.js
+++ b/tests/rules/no-invalid-properties.test.js
@@ -295,6 +295,26 @@ ruleTester.run("no-invalid-properties", rule, {
 				},
 			],
 		},
+
+		/*
+		 * Reporting an unknown `var()` nested inside `env()` is intentional:
+		 * this rule also checks that custom properties are resolvable, and that
+		 * check is independent of the syntax validation skipped for `env()`.
+		 * We can revisit this once the rule supports partial validation of values.
+		 */
+		{
+			code: "a { padding: env(safe-area-inset-top, var(--external-padding)); }",
+			errors: [
+				{
+					messageId: "unknownVar",
+					data: { var: "--external-padding" },
+					line: 1,
+					column: 43,
+					endLine: 1,
+					endColumn: 61,
+				},
+			],
+		},
 		{
 			code: "a { color: bar }",
 			errors: [

